pkg/broker/handler/processors/fanout/processor.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fanout
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/cloudevents/sdk-go/v2/event"
+	"go.uber.org/zap"
+	"knative.dev/eventing/pkg/logging"
+
+	"github.com/google/knative-gcp/pkg/broker/config"
+	handlerctx "github.com/google/knative-gcp/pkg/broker/handler/context"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors"
+)
+
+type fanoutResult struct {
+	err error
+}
+
+// Processor fanouts an event based on the broker key in the context.
+type Processor struct {
+	processors.BaseProcessor
+
+	// MaxConcurrency is the max number of goroutine it will spawn
+	// for each event.
+	MaxConcurrency int
+
+	// Targets is the targets from config.
+	Targets config.ReadonlyTargets
+}
+
+// Process fanouts the given event.
+func (p *Processor) Process(ctx context.Context, event *event.Event) error {
+	bk := handlerctx.GetBrokerKey(ctx)
+	broker, ok := p.Targets.GetBrokerByKey(bk)
+	if !ok {
+		return fmt.Errorf("event broker %q doesn't exist in config", bk)
+	}
+	ctx = handlerctx.WithBroker(ctx, broker)
+
+	tc := make(chan *config.Target)
+	go func() {
+		defer close(tc)
+		for _, target := range broker.Targets {
+			tc <- target
+		}
+	}()
+
+	curr := len(broker.Targets)
+	if curr > p.MaxConcurrency {
+		curr = p.MaxConcurrency
+	}
+
+	var resChs []<-chan *fanoutResult
+	for i := 0; i < curr; i++ {
+		resChs = append(resChs, p.fanoutEvent(ctx, event, tc))
+	}
+
+	return p.mergeResults(ctx, resChs)
+}
+
+func (p *Processor) fanoutEvent(ctx context.Context, event *event.Event, tc <-chan *config.Target) <-chan *fanoutResult {
+	out := make(chan *fanoutResult)
+	go func() {
+		defer close(out)
+		for target := range tc {
+			ctx = handlerctx.WithTarget(ctx, target)
+			out <- &fanoutResult{err: p.Next().Process(ctx, event)}
+		}
+	}()
+	return out
+}
+
+func (p *Processor) mergeResults(ctx context.Context, resChs []<-chan *fanoutResult) error {
```

Comment:
Not sure if knative has similar utilities but k8s has a workqueue utility that helps parallelize workers:

https://github.com/kubernetes/kubernetes/blob/a10021787b2d037e3c1d44a205980fd48ac23eff/staging/src/k8s.io/client-go/util/workqueue/parallelizer.go#L30